### ISSUE TITLE
About data preprocessing

### ISSUE THREAD

**Issue by zmyjoe** [yzz673/Brant#1]
Hello! Thank you for this amazing work! I'm trying to finetune Brant on some datasets including MAYO and FNUSA, but there're some unclear details on preprocessing, such as how to organize the patches into sequences and how to organize MAYO/FNUSA's channels. Could you please offer the preprocessing code or describe the details?

**Comment by yzz673** [comment]
> Hello! Thank you for this amazing work! I'm trying to finetune Brant on some datasets including MAYO and FNUSA, but there're some unclear details on preprocessing, such as how to organize the patches into sequences and how to organize MAYO/FNUSA's channels. Could you please offer the preprocessing code or describe the details?

Hello! Thank you for your interest in our work! Brant contains 2 parts: the Time encoder and the Channel encoder. If your datasets are shaped like [channel x data], then you should segment your data to segments with shape [channel x 15 x 1500] and finetune both the Time encoder and Channel encoder. If your datasets are like MAYO/FNUSA which have already been processed into segments with shape [15000], then you should treat each segment as a sample(sequence). In our preprocessing, we first upsample each segment from 15000 to 22500 by `np.interp`. Then we reshape each segment to [15, 1500] and only finetune the Time encoder of Brant with an additional MLP classifier.

**Comment by zmyjoe** [comment]
Thank you for your reply! Another question: how do I apply MAYO/FNUSA to prediction task as predicting requires longer sequence length (history + prediction)?

**Comment by DaozeZhang** [comment]
> Thank you for your reply! Another question: how do I apply MAYO/FNUSA to prediction task as predicting requires longer sequence length (history + prediction)?

In order to perform forecasting tasks on the MAYO and FNUSA datasets, considering that the input shape of the model is (15, 1500), we designed an implementation that is equivalent to the forecasting task. Taking `history_len = 12` and  `forecast_len = 3` segments as an example, for signals with length of 22500, we mask the last 3*1500 points (that is, the last 3 segments), and then let the model reconstruct these 3 segments. In this way, the mask-reconstruction task of the model can be used to achieve the effect of the forecasting task.

**Comment by rasgaard** [comment]
Just chiming in to say that I have to agree with @zmyjoe about providing code for preprocessing. With such great work it would be a real shame for the reproducability and applicability to suffer due to a missing link in the data processing pipeline. 

I applaud the authors for taking the time to answer issues here on GitHub - it really is greatly appreciated - but as we all know some details are only revealed by reading through the actual code.

**Comment by yzz673** [comment]
We upload the preprocessing code for MAYO and FNUSA datasets in `Brant_src/mayo_fnusa_preprocess.py`.

**Comment by rasgaard** [comment]
> We upload the preprocessing code for MAYO and FNUSA datasets in `Brant_src/mayo_fnusa_preprocess.py`.

Thanks for this! 

I'm wondering if there are separate utility scripts for the MAYO and FNUSA datasets? 

My suspicion is due to the preprocessing in `Brant_src/mayo_fnusa_preprocess.py` produces segment-wise files as well as one big aggregated file while functions like `_load_data` in `Brant_src/utils.py` seems to expect files on a patient-level. Perhaps I have missed some patient-level aggregation steps?

> we designed an implementation that is equivalent to the forecasting task.

Are there are also separate training scripts for MAYO and FNUSA?

**Comment by DaozeZhang** [comment]
> > We upload the preprocessing code for MAYO and FNUSA datasets in `Brant_src/mayo_fnusa_preprocess.py`.
> 
> Thanks for this!
> 
> I'm wondering if there are separate utility scripts for the MAYO and FNUSA datasets?
> 
> My suspicion is due to the preprocessing in `Brant_src/mayo_fnusa_preprocess.py` produces segment-wise files as well as one big aggregated file while functions like `_load_data` in `Brant_src/utils.py` seems to expect files on a patient-level. Perhaps I have missed some patient-level aggregation steps?
> 
> > we designed an implementation that is equivalent to the forecasting task.
> 
> Are there are also separate training scripts for MAYO and FNUSA?

According to Appendix G in the paper, we take the model that was fine-tuned on our own data and do inference on MAYO and FNUSA datasets. Therefore, there is no training script on these two public datasets.

**Comment by rasgaard** [comment]
> According to Appendix G in the paper, we take the model that was fine-tuned on our own data and do inference on MAYO and FNUSA datasets. Therefore, there is no training script on these two public datasets.

Thanks for elaborating!

I still feel like some aspects related to the data preprocessing is unclear though, 

`Brant_src/mayo_fnusa_preprocess.py` produces files in the form of `[segment_id]_[15*1500].npy` (for example, `x000000_22500.npy`) 
https://github.com/yzz673/Brant/blob/ea61462fbc994b0cfac227a48baaa7db9f5ff125/Brant_src/mayo_fnusa_preprocess.py#L36

However, if I attempt to run any evaluation scripts through `main.py` I will need to load in the data, which is done on a patient level which can be seen in `Brant_src/utils.py`
https://github.com/yzz673/Brant/blob/ea61462fbc994b0cfac227a48baaa7db9f5ff125/Brant_src/utils.py#L33

Is there an aggregation step from segment to patient-level data files?

The only aggregation I see is seemingly unused. What is the purpose of this file?

https://github.com/yzz673/Brant/blob/ea61462fbc994b0cfac227a48baaa7db9f5ff125/Brant_src/mayo_fnusa_preprocess.py#L50


Thank you for engaging in discussions about the code. It is a very interesting project! :)

**Comment by DaozeZhang** [comment]
On the MAYO and FNUSA datasets, there is no aggregateion between channels. These two datasets do not provide any degree of channel information, so the channel dimension can only be `1` in each data segment.

**Comment by TousenKaname** [comment]
> Hello! Thank you for this amazing work! I'm trying to finetune Brant on some datasets including MAYO and FNUSA, but there're some unclear details on preprocessing, such as how to organize the patches into sequences and how to organize MAYO/FNUSA's channels. Could you please offer the preprocessing code or describe the details?

Hi! Did you find MAYO and FUNSA dataset?  Could you please share the link where to download these two datasets? :)  I tried to search in google but can't find out the links...
